jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala

```diff
@@ -257,21 +266,19 @@ object XGBoost extends Serializable {
    * @param useExternalMemory indicate whether to use external memory cache, by setting this flag as
    *                          true, the user may save the RAM cost for running XGBoost within Spark
    * @param missing the value represented the missing value in the dataset
-   * @param baseMargin initial prediction for boosting.
    * @throws ml.dmlc.xgboost4j.java.XGBoostError when the model training is failed
    * @return XGBoostModel when successful training
    */
   @throws(classOf[XGBoostError])
   def trainWithRDD(
-      trainingData: RDD[MLLabeledPoint],
+      trainingData: RDD[XGBLabeledPoint],
```

Comment:
`zip` doesn't work because it has nontrivial assumptions about the RDDs being zipped, [namely](https://spark.apache.org/docs/2.2.0/api/java/org/apache/spark/rdd/RDD.html#zip-org.apache.spark.rdd.RDD-scala.reflect.ClassTag-)

> Assumes that the two RDDs have the *same number of partitions* and the *same number of elements in each partition* (e.g. one was made through a map on the other).

Note that for our purpose not only the number of elements in each partition should match between the two RDDs but also their ordering. In practice, these assumptions might hard to satisfy, e.g. if the RDDs are completely independent -- you read the training data from one file and the baseMargin from another one.

Maybe you can get it to work by using a specific partitioner with `repartitionAndSortWithinPartitions`, but other than that Spark does not offer any API to align partitions of one RDD with another one. This looks like a problem to me since we would need to propagate this assumption to our APIs.